### PR TITLE
Remove the NetworkChan

### PR DESCRIPTION
**Pull request by tomaka**
- `protocol.rs` now communicates back its intentions towards the network through a `NetworkOut` trait that is passed whenever a method is called.
- Removes the `NetworkChan` type altogether, and replaces it internally with an `UnboundedSender` from the `futures` crate.

After that, the codes of `Protocol` and `Service` are much already much cleaner in my opinion. There are a lot of small adjustments to make that would be off-topic for this PR, so I'll do them on top.

This PR really makes the code of the test helpers much more dirty (it already was), but the next step will be to rework that.

### CODE REVIEW

File: core/network/src/service.rs [hunk 1]
```diff
@@ -688,71 +625,78 @@ fn run_thread<B: BlockT + 'static, S: NetworkSpecialization<B>, H: ExHashT>(
...
 		loop {
-			let outcome = match network_service.lock().poll() {
+			let mut network_service = network_service.lock();
+			let poll_value = network_service.poll();
+			let mut network_out = Ctxt(&mut network_service, &peerset);
+
+			let outcome = match poll_value {
 				Ok(Async::NotReady) => break,
 				Ok(Async::Ready(Some(NetworkServiceEvent::OpenedCustomProtocol { peer_id, version, debug_info, .. }))) => {
 					debug_assert!(
```
[the remark this answers is not in the document]
Comment (tomaka) [reply core/network/src/service.rs]: We're kind of doing defensive programming. We don't want release builds to panic, ever, but debug_asserts are fine.

**Comment by tomaka** [comment]
The CI is failing because it turns out that one of the aura tests uses the network testnet helper from within a tokio runtime.

I will have to do some cleanup in the network tests for this to work.

**Comment by tomaka** [comment]
1. For the libp2p service and the protocol in particular, the answer is yes. The libp2p service does almost nothing. It reads events from a channel (that's in libp2p's code), eventually updates some local state (libp2p and network-libp2p), then sends messages to other channels (in network::service.rs). Its CPU usage is minimal, and to me it doesn't really make sense to me to run this as its own thread.

However it is clearly not my intention to make them impossible to split.
My intention is that instead of having, say, structs `A`, `B` and `C` that expose an asynchronous API, you would split each of them between `Async`, `Aasync`, `Bsync`, `Basync`, and `Csync`, `CAsync`. Then you leave `Async`, `Bsync` and `Csync` in their own module with their synchronous API, and you move `Aasync`, `Basync` and `Casync` to a single module and handle all the asynchronicity there.
The logic would remain in `Async`, `Bsync` and `Csync`, while the additional complexity introduced by concurrency would be contained in a single module that would be responsible for figuring out things such as ordering.

That's what I'm doing for `protocol.rs`: I'm basically splitting it in two: `ProtocolSync` (whose name stays `Protocol`), and `ProtocolAsync` (which is unnamed in the code but more or less consists in the `run_thread` function) that resides in `service.rs`.

Another big advantage of this, which is a bit tricky to illustrate, is that communications have a more precise meaning. By having services directly communicate to each other, you sometimes have to pass some context which would be out of scope of any of the two services individually.
For example right now the import queue API has an `import_block` method. If you move all the asynchronicity at the same place, you can easily know which blocks came from the networking and which blocks came from mining, and can for example prioritize the blocks coming from the mining. Contrast this with having to pass some sort of "priority level value" to the networking and the mining codes.

2. Yes, it is indeed my intention to entirely remove `run_thread` and move what it's doing to a `poll()` method that could spawn background tasks. That's more of a long term objective for #2416 .

3. Yes I've noticed that the ordering could potentially change here. In practice this only concerns the `OnDemandService`, as it is the only object that still indirectly uses the channel. Its requests could be delayed a bit compared to now. In practice it doesn't matter anyway, because: 1- on the network we only send requests that are all independent of each other, 2- the `OnDemandService` is running in a different thread than the networking, and therefore ordering wasn't guaranteed before anyway.

However note that this kind of tricky situations is exactly what I want to avoid by reworking the concurrency story of the network. I totally agree that it can indeed lead to subtle bugs, and more determinism not only makes things more simple to reason about, but also makes it easier to detect and reproduce bugs.

> For example, do we really want all the stuff happening in ChainSync to always fully block networking, and vice-versa?

If something expensive is happening in `ChainSync`, then that specific thing should be made asynchronous and not `ChainSync` as a whole.
For example, right now importing a chunk of blocks can take up to 30 seconds in debug mode (for a reason I have yet to examine). If everything was put into a single task/thread, then indeed this would block the entire networking for 30 seconds.
However, note that even if `ChainSync` is in its own dedicated task, importing a chunk of blocks would still block any incoming request for 30 seconds. The solution to me is to specifically extract the thing that takes a lot of CPU, and not to make the entire `ChainSync` in its own task.

> In any case, I'd like to see an explicit discussion about how we intend to "run" protocol, networking, and the other components communicating with them, that addresses the concurrency of the system(or lack thereof). Otherwise we run the risk of doing everything in a single task, which in the immediate might appear "simpler", and then later realizing that we'd like to get some parallelism between various operations, which would then be much harder to achieve.

First of all, if we have to choose between "simple and one task" and "hard and multithreaded", I would argue that we should lean towards "hard and multithreaded" only if there's a benefit to multithreading things. Otherwise we run into premature optimizations. I clearly think that we can "asyncify" things later on top of synchronous API (with the `Async`/`Aasync` splitting thingy), without having to take decisions too far ahead.

As for a precise answer, the way I see it in the very long term is:

- Each TCP connection has its own dedicated task (that's already the case in libp2p's code, but hidden beneath the API) that handle parsing networking messages.
- The protocol/network task gathers messages received by these per-connection tasks, handles all the logic related to protocols (ie. if we receive an answer), and generates messages for the consensus, the import queue, and requests for the client. This would also contain the on-demand service.
- The client has its own task, and asynchronously handles all database read/write requests. For example it would read a request for the headers of blocks 8022 to 8290, fetch them from its database, and answer. For the light client, it optionally generates network requests itself that are handled by the protocol/network task. Right now this is already more or less the case, except that it's done by locking mutexes instead of sending messages.
- The import queue has its own task, and is between the protocol/network and the client, as it is already the case.

**Comment by tomaka** [comment]
Fixing the network tests is hell on earth. Half of the implementation details have public visibility.

**Comment by gterzian** [comment]
Thanks for providing additional info, overall I think it sounds like a reasonable approach, and I do have a few more comments:

> I'm basically splitting it in two: ProtocolSync (whose name stays Protocol), and ProtocolAsync (which is unnamed in the code but more or less consists in the run_thread function) that resides in service.rs.

I can support the goal of having simple sync api, with some sort of wrapper around them to make them async where needed, and the only thing I am worried about is that if making things "async" means having to share entire stucts like `protocol` across threads, we will be back to having comments everywhere about "take the locks in that order" and so on(and few operations will be really async, due to the synchronization requirements). 

For example if I now look at `run_thread`, I think that making the various operations taking place async, as opposed to all in a single task, would require sharing state across threads(or tasks), because the API of the various components are now method based, versus message based. Except off-course if you were to re-introduce channels and call the "sync" api in response to receiving "async" messages inside some sort of wrapper. And wouldn't that be sort-of re-introducing what has been removed in this PR?

> If something expensive is happening in ChainSync, then that specific thing should be made asynchronous and not ChainSync as a whole.

I'm not sure if there is anything really "expensive" going in in ChainSync, or other parts of protocol like gossip, yet I do think quite a few of these operations could probably be done while the network service does other things. However splitting those off separately from Protocol as a whole is probably not worth it since it introduce complications around sharing the state necessary to perform those. 

That's essentially while I though having `protocol` run in a thread, or maybe rather a task, as a whole and separate from the actual network service, was a pretty decent idea.

**Comment by gterzian** [comment]
Ok, thanks for the explanations and looking forward to see where this will lead...

**Comment by tomaka** [comment]
Ready for review.
I initially had the intention to rework the network tests, but this task is still way too huge at the moment.
Instead, I added a `uses_tokio() -> bool` method on `TestNetFactory`.

File: core/network/src/test/mod.rs [hunk 2]
```diff
@@ -377,7 +403,20 @@ impl<S: NetworkSpecialization<Block>> ProtocolChannel<S> {
...
+				Ok(Async::NotReady) => None,
+				Err(_) => None,
+				Ok(Async::Ready(None)) => None,
+			}))
+		});
+	
+		if self.use_tokio {
+			fut.wait()
+		} else {
+			tokio::runtime::current_thread::block_on_all(fut)
```
[the remark this answers is not in the document]
Comment (tomaka) [reply core/network/src/test/mod.rs]: The `block_on_all` function always creates a new runtime, which results in a panic if we are already within a runtime. This is why I had to introduce this `use_tokio` thing.

[the remark this answers is not in the document]
Comment (gterzian) [reply core/network/src/test/mod.rs]: Ok, thanks for the info...

**Pull request merged**